File: libviews/src/main/java/com/uncle2000/libviews/TitleView.java
package com.uncle2000.libviews;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;


/**
 * Created by xiao on 2016/6/21.
 * 功能:
 */
public class TitleView extends FrameLayout implements View.OnClickListener {


    private static final String TAG = "TitleView";

    protected TextView title_tv, title_right_tv, title_left_tv;

    protected TextView tvSubTitle;
    protected ImageView title_left_iv, title_right_iv, title_right_iv2;

    private View title_left_fl, title_right_fl;

    private boolean needFitStatusBar = true;
    private boolean haveFitStatusBar = false;

    public TitleView(Context context) {
        super(context);
        init(context, null);
    }

    public TitleView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    /**
     * It's good for viewgroup,not for view.U can put view in framelayout.
     */
    public static void setViewPadding(Context context, View... views) {

        //        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
        //            return;
        //        }
        //        if (Configure.barHeight == 0) {
        //            new getStatusBarHeight().getStatusBarHeight(context);
        //        }
        //        //		if(null==views || views.length==0){
        //        //
        //        //		}
        //        for (View v : views) {
        //            if (null == v)
        //                continue;
        //            v.setPadding(0, Configure.barHeight, 0, 0);
        //        }
    }

    private void init(Context context, AttributeSet attrs) {
        LayoutInflater.from(context).inflate(R.layout.view_title_bar, this);

        title_tv = findViewById(R.id.title_tv);
        tvSubTitle = findViewById(R.id.sub_title_tv);
        title_left_iv = findViewById(R.id.title_left_iv);
        title_right_iv = findViewById(R.id.title_right_iv);
        title_right_tv = findViewById(R.id.title_right_tv);
        title_left_tv = findViewById(R.id.title_left_tv);
        title_right_iv2 = findViewById(R.id.title_right_iv2);
        title_left_fl = findViewById(R.id.title_left_fl);
        title_right_fl = findViewById(R.id.title_right_fl);

        if (attrs != null) {
            TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.TitleView);

            String leftText = null; //  = a.getAny(R.styleable.TitleView_leftText);
            String rightText = null; // = a.getAny(R.styleable.TitleView_rightText);
            String titleText = null; //  = a.getAny(R.styleable.TitleView_titleText);
            String subTitleText = null; // = a.getAny(R.styleable.TitleView_subTitleText);
            Drawable leftDrawable = ContextCompat.getDrawable(getContext(), R.drawable.svg_arrow_left); // = a.getDrawable(R.styleable.TitleView_leftDrawable);
            Drawable rightDrawable = null; // = a.getDrawable(R.styleable.TitleView_rightDrawable);
            Drawable rightDrawable2 = null; // = a.getDrawable(R.styleable.TitleView_rightDrawable);

            boolean showDivider = true;
            final int N = a.getIndexCount();
            for (int i = 0; i < N; i++) {
                int attr = a.getIndex(i);
                if (attr == R.styleable.TitleView_leftText) {
                    leftText = a.getString(attr);

                } else if (attr == R.styleable.TitleView_rightText) {
                    rightText = a.getString(attr);

                } else if (attr == R.styleable.TitleView_titleText) {
                    titleText = a.getString(attr);

                } else if (attr == R.styleable.TitleView_subTitleText) {
                    subTitleText = a.getString(attr);

                } else if (attr == R.styleable.TitleView_leftDrawable) {
                    leftDrawable = a.getDrawable(attr);

                } else if (attr == R.styleable.TitleView_rightDrawable) {
                    rightDrawable = a.getDrawable(attr);

                } else if (attr == R.styleable.TitleView_rightDrawable2) {
                    rightDrawable2 = a.getDrawable(attr);

                } else if (attr == R.styleable.TitleView_showBottomDivider) {
                    showDivider = a.getBoolean(attr, showDivider);
                } else if (attr == R.styleable.TitleView_needFitStatusBar) {
                    needFitStatusBar = a.getBoolean(attr, needFitStatusBar);
                }
            }
            a.recycle();

            setLeftText(leftText);
            setRightText(rightText);
            setTitleText(titleText);
            setImageDrawable(title_left_iv, leftDrawable);
            setImageDrawable(title_right_iv, rightDrawable);
            setImageDrawable(title_right_iv2, rightDrawable2);

            if (!showDivider) findViewById(R.id.divider).setVisibility(View.GONE);
            setSubTitle(subTitleText);

        }

        // ClickableImageTouchListener.addTouchDrak(title_left_iv);
        setLeftOnClickListener(this);

        setViewPadding(getContext(), this);

        // boolean fit = false;
        // if (Build.VERSION.SDK_INT > 16) {
        //     fit = getFitsSystemWindows();
        // } else {
        //     // fit = fitsSystemWindows();
        // }
        // if (fit)
        // setPadding(0, StatusBarTool.getStatusBarHeight(getContext()), 0, 0);
    }

    private void setImageDrawable(ImageView iv, Drawable drawable) {
        if (drawable != null) {
            iv.setImageDrawable(drawable);
            iv.setVisibility(VISIBLE);
        } else {
            iv.setVisibility(View.GONE);
        }
    }

    public void setLeftText(String text) {
        if (TextUtils.isEmpty(text)) {
            title_left_tv.setVisibility(GONE);
            return;
        }
        title_left_tv.setVisibility(VISIBLE);
        title_left_tv.setText(text);
    }

    public void setLeftText(int res) {
        if (res <= 0) {
            title_left_tv.setVisibility(GONE);
            return;
        }
        title_left_tv.setVisibility(VISIBLE);
        title_left_tv.setText(res);
    }

    public void setRightText(String text) {
        if (TextUtils.isEmpty(text)) {
            title_right_tv.setVisibility(GONE);
            return;
        }

        title_right_tv.setVisibility(VISIBLE);
        title_right_tv.setText(text);
    }

    public void setRightText(int res) {
        if (res <= 0) {
            title_right_tv.setVisibility(GONE);
            return;
        }
        title_right_tv.setVisibility(VISIBLE);
        title_right_tv.setText(res);
    }

    public void setTitleText(String text) {
        if (text != null) {
            title_tv.setText(text);
        } else {
            title_tv.setText("");
        }
    }

    public void setTitleText(int res) {
        title_tv.setText(res);
    }

    public void setSubTitle(int res) {
        if (res <= 0) {
            tvSubTitle.setVisibility(View.GONE);
        } else {
            tvSubTitle.setVisibility(View.VISIBLE);
            tvSubTitle.setText(res);
        }
    }

    public void setSubTitle(CharSequence txt) {
        if (TextUtils.isEmpty(txt)) {
            tvSubTitle.setVisibility(View.GONE);
        } else {
            tvSubTitle.setVisibility(View.VISIBLE);
            tvSubTitle.setText(txt);
        }
    }

    public void setLeftDrawable(int res) {
        if (res <= 0) {
            title_left_iv.setVisibility(View.GONE);
            title_left_iv.setImageDrawable(null);
        } else {
            title_left_iv.setVisibility(View.VISIBLE);
            title_left_iv.setImageResource(res);
        }
    }

    public void setRightDrawable(int res) {
        if (res > 0) {
            title_right_iv.setVisibility(View.VISIBLE);
            title_right_iv.setImageResource(res);
        } else {
            title_right_iv.setVisibility(View.GONE);
        }
    }

    public void setLeftDrawableVisible(boolean visible) {
        if (visible) {
            title_left_iv.setVisibility(VISIBLE);
        } else {
            title_left_iv.setVisibility(GONE);
        }
    }

    public void setRightDrawableVisible(boolean visible) {
        if (visible) {
            title_right_iv.setVisibility(VISIBLE);
        } else {
            title_right_iv.setVisibility(INVISIBLE);
        }
    }

    public void setRightTextVisible(boolean visible) {
        if (visible) {
            title_right_tv.setVisibility(VISIBLE);
        } else {
            title_right_tv.setVisibility(GONE);
        }
    }

    public void setLeftOnClickListener(OnClickListener listener) {
        title_left_fl.setOnClickListener(listener);
    }

    public void setRightOnClickListener(OnClickListener listener) {
        title_right_fl.setOnClickListener(listener);
    }

    public void setRightOnClickListener2(OnClickListener listener) {
        title_right_iv2.setOnClickListener(listener);
    }


    @Override
    public void onClick(View v) {

        if (v == title_left_fl) {
            // 内容已经全部被隐藏
            if (title_left_iv.getVisibility() != VISIBLE && title_left_tv.getVisibility() != VISIBLE) {
                return;
            }
            if (getContext() instanceof Activity) {
                ((Activity) getContext()).onBackPressed();
            }
        }
    }

    public void setRightIcon2(int i) {
        if (i <= 0) {
            title_right_iv2.setVisibility(View.GONE);
        } else {
            title_right_iv2.setVisibility(View.VISIBLE);
            title_right_iv2.setImageResource(i);
        }
    }

    public void setRightIcon2Visible(boolean visible) {
        title_right_iv2.setVisibility(visible ? VISIBLE : GONE);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);

//        if (needFitStatusBar) {
//            int statusBarHeight = StatusBarTool.getStatusBarHeight(getContext());
//            int[] loc = new int[2];
//            getLocationOnScreen(loc);
//            int y = loc[1];
//
//            int requiredPaddingTop = y == 0 ? statusBarHeight : 0;
//            if (getPaddingTop() != requiredPaddingTop) {
//                post(() -> {
//                    setPadding(getPaddingLeft(), requiredPaddingTop, getPaddingRight(), getPaddingBottom());
//
//                });
//            }
//        }
    }
}

File: libviews/src/main/java/com/uncle2000/libviews/TitleView.kt
package com.uncle2000.libviews

import android.app.Activity
import android.content.Context
import android.databinding.DataBindingUtil
import android.graphics.drawable.Drawable
import android.support.constraint.ConstraintLayout
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import com.blankj.utilcode.util.ConvertUtils
import com.qmuiteam.qmui.util.QMUIStatusBarHelper
import com.uncle2000.libviews.databinding.ViewTitleBinding

class TitleView @JvmOverloads constructor(
        context: Context,
        private val attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : ConstraintLayout(context, attrs, defStyleAttr), View.OnClickListener {
    val TYPE_NONE = 0
    val TYPE_WHITE = 1
    val TYPE_BLACK = 2

    var leftDrawable = ContextCompat.getDrawable(context, R.drawable.svg_arrow_left_black)
    var rightDrawable: Drawable? = null
    var right2Drawable: Drawable? = null
    var leftText = ""
    var rightText = ""
    var right2Text = ""
    var titleText = ""
    var subTitleText = ""
    var titleHeight: Int = ConvertUtils.dp2px(48f)
    var leftTextSize: Int = ConvertUtils.dp2px(14f)
    var rightTextSize: Int = ConvertUtils.dp2px(14f)
    var right2TextSize: Int = ConvertUtils.dp2px(14f)
    var titleTextSize: Int = ConvertUtils.dp2px(16f)
    var subTitleTextSize: Int = ConvertUtils.dp2px(15f)
    var leftTextColor = ContextCompat.getColor(context, R.color.defaults_title_text_color)
    var rightTextColor = ContextCompat.getColor(context, R.color.defaults_other_title_text_color)
    var right2TextColor = ContextCompat.getColor(context, R.color.defaults_other_title_text_color)
    var titleTextColor = ContextCompat.getColor(context, R.color.defaults_other_title_text_color)
    var subTitleTextColor = ContextCompat.getColor(context, R.color.defaults_sub_title_text_color)
    //    var backgroundColor: Int = 0
    var dividerColor = ContextCompat.getColor(context, R.color.defaults_divider_color)
    var showBottomDivider = false
    var needFitStatusBar = true
    var colorMode = TYPE_NONE
    var leftVisibility = true
    var rightVisibility = false
    var right2Visibility = false
    var titlePaddingTop = 10

    private val binding: ViewTitleBinding =
            DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.view_title, this, true)

    init {
        setLayerType(View.LAYER_TYPE_SOFTWARE, null)
        /*放到这里 以便设置编辑框前可以得到参数*/
        initAttrs()

        setLeftOnClickListener(this)
    }

    private fun initAttrs() {
        if (attrs != null) {
            val a = context.obtainStyledAttributes(attrs, R.styleable.TitleView)

            for (i in 0 until a.indexCount) {
                val attr = a.getIndex(i)
                when (attr) {
                    R.styleable.TitleView_colorMode -> colorMode = a.getInt(attr, colorMode)
                    R.styleable.TitleView_needFitStatusBar -> needFitStatusBar = a.getBoolean(attr, true)
                    R.styleable.TitleView_titleHeight -> titleHeight = a.getDimension(attr, titleHeight.toFloat()).toInt()
                    R.styleable.TitleView_titlePaddingTop -> titlePaddingTop = a.getDimension(attr, titlePaddingTop.toFloat()).toInt()
                }
            }
            binding.apply {
                if (needFitStatusBar) {
                    val statusBarHeight = QMUIStatusBarHelper.getStatusbarHeight(context)
                    if (statusBarHeight >= 0 && titlePaddingTop >= 0) {
                        QMUIStatusBarHelper.translucent(context as Activity)
                        paddingTopView.layoutParams = ConstraintLayout.LayoutParams(
                                LayoutParams.MATCH_PARENT, titlePaddingTop + statusBarHeight)
                    }
                    root.layoutParams = ConstraintLayout.LayoutParams(LayoutParams.MATCH_PARENT, titleHeight + titlePaddingTop + statusBarHeight)
                } else {
                    if (titlePaddingTop >= 0)
                        paddingTopView.layoutParams = ConstraintLayout.LayoutParams(
                                LayoutParams.MATCH_PARENT, titlePaddingTop)
                    root.layoutParams = ConstraintLayout.LayoutParams(LayoutParams.MATCH_PARENT, titleHeight + titlePaddingTop)
                }
                when (colorMode) {
                    TYPE_WHITE -> {
                        if (needFitStatusBar)
                            QMUIStatusBarHelper.setStatusBarDarkMode(context as Activity)
                        leftDrawable = ContextCompat.getDrawable(context, R.drawable.svg_arrow_left_white)
                        leftTextColor = ContextCompat.getColor(context, R.color.defaults_white_text_color)
                        rightTextColor = ContextCompat.getColor(context, R.color.defaults_white_text_color)
                        right2TextColor = ContextCompat.getColor(context, R.color.defaults_white_text_color)
                        titleTextColor = ContextCompat.getColor(context, R.color.defaults_white_text_color)
                        subTitleTextColor = ContextCompat.getColor(context, R.color.defaults_white_text_color)
                    }
                    TYPE_BLACK -> {
                        if (needFitStatusBar)
                            QMUIStatusBarHelper.setStatusBarLightMode(context as Activity)
                        leftDrawable = ContextCompat.getDrawable(context, R.drawable.svg_arrow_left_black)
                        leftTextColor = ContextCompat.getColor(context, R.color.defaults_title_text_color)
                        rightTextColor = ContextCompat.getColor(context, R.color.defaults_other_title_text_color)
                        right2TextColor = ContextCompat.getColor(context, R.color.defaults_other_title_text_color)
                        titleTextColor = ContextCompat.getColor(context, R.color.defaults_other_title_text_color)
                        subTitleTextColor = ContextCompat.getColor(context, R.color.defaults_sub_title_text_color)
                    }
                    else -> {
                    }
                }
            }
            for (i in 0 until a.indexCount) {
                val attr = a.getIndex(i)
                when (attr) {
                    R.styleable.TitleView_leftDrawable -> leftDrawable = a.getDrawable(attr)
                    R.styleable.TitleView_rightDrawable -> rightDrawable = a.getDrawable(attr)
                    R.styleable.TitleView_rightDrawable2 -> right2Drawable = a.getDrawable(attr)
                    R.styleable.TitleView_leftText -> leftText = a.getString(attr)
                    R.styleable.TitleView_rightText -> rightText = a.getString(attr)
                    R.styleable.TitleView_right2Text -> right2Text = a.getString(attr)
                    R.styleable.TitleView_titleText -> titleText = a.getString(attr)
                    R.styleable.TitleView_subTitleText -> subTitleText = a.getString(attr)
                    R.styleable.TitleView_leftTextSize -> leftTextSize = a.getDimensionPixelSize(attr, leftTextSize)
                    R.styleable.TitleView_rightTextSize -> rightTextSize = a.getDimensionPixelSize(attr, rightTextSize)
                    R.styleable.TitleView_right2TextSize -> right2TextSize = a.getDimensionPixelSize(attr, right2TextSize)
                    R.styleable.TitleView_titleTextSize -> titleTextSize = a.getDimensionPixelSize(attr, titleTextSize)
                    R.styleable.TitleView_subTitleTextSize -> subTitleTextSize = a.getDimensionPixelSize(attr, subTitleTextSize)
                    R.styleable.TitleView_leftTextColor -> leftTextColor = a.getColor(attr, leftTextColor)
                    R.styleable.TitleView_rightTextColor -> rightTextColor = a.getColor(attr, rightTextColor)
                    R.styleable.TitleView_right2TextColor -> right2TextColor = a.getColor(attr, right2TextColor)
                    R.styleable.TitleView_titleTextColor -> titleTextColor = a.getColor(attr, titleTextColor)
                    R.styleable.TitleView_subTitleTextColor -> subTitleTextColor = a.getColor(attr, subTitleTextColor)
//                    R.styleable.TitleView_backgroundColor -> backgroundColor = a.getColor(attr, )
                    R.styleable.TitleView_dividerColor -> dividerColor = a.getColor(attr, dividerColor)
                    R.styleable.TitleView_showBottomDivider -> showBottomDivider = a.getBoolean(attr, false)
                    R.styleable.TitleView_leftVisibility -> leftVisibility = a.getBoolean(attr, true)
                    R.styleable.TitleView_rightVisibility -> rightVisibility = a.getBoolean(attr, false)
                    R.styleable.TitleView_right2Visibility -> right2Visibility = a.getBoolean(attr, false)
                }
            }
            a.recycle()
        }

        binding.apply {
            titleLeft.visibility = if (leftVisibility)
                View.VISIBLE
            else
                View.GONE
            titleRight.visibility = if (rightVisibility)
                View.VISIBLE
            else
                View.GONE
            titleRight2.visibility = if (right2Visibility)
                View.VISIBLE
            else
                View.GONE
            titleLeftIv.setImageDrawable(leftDrawable)
            titleRightIv.setImageDrawable(rightDrawable)
            titleRightIv2.setImageDrawable(right2Drawable)
            titleLeftTv.text = leftText
            titleLeftTv.setTextColor(leftTextColor)
            titleLeftTv.setTextSize(TypedValue.COMPLEX_UNIT_PX, leftTextSize.toFloat())
            titleRightTv.text = rightText
            titleRightTv.setTextColor(rightTextColor)
            titleRightTv.setTextSize(TypedValue.COMPLEX_UNIT_PX, rightTextSize.toFloat())
            titleRightTv2.text = right2Text
            titleRightTv2.setTextColor(right2TextColor)
            titleRightTv2.setTextSize(TypedValue.COMPLEX_UNIT_PX, right2TextSize.toFloat())
            title.text = titleText
            title.setTextColor(titleTextColor)
            title.setTextSize(TypedValue.COMPLEX_UNIT_PX, titleTextSize.toFloat())
            subTitle.text = subTitleText
            subTitle.setTextColor(subTitleTextColor)
            subTitle.setTextSize(TypedValue.COMPLEX_UNIT_PX, subTitleTextSize.toFloat())
            divider.setBackgroundColor(dividerColor)
            divider.visibility = if (showBottomDivider)
                View.VISIBLE
            else
                View.GONE
        }
    }

    override fun onClick(v: View) {
        if (v.id == R.id.title_left) {
            // 内容已经全部被隐藏
            if (binding.titleLeft.visibility != View.VISIBLE && binding.titleLeftTv.visibility != View.VISIBLE) {
                return
            }
            if (context is Activity) {
                (context as Activity).onBackPressed()
            }
        }
    }

    fun setLeftOnClickListener(listener: View.OnClickListener) {
        binding.titleLeft.setOnClickListener(listener)
    }

    @JvmOverloads
    fun setTitleContent(titleText: String = "", leftTitleText: String = "", rightTitleText: String = "", subTitleText: String = "") {
        if (titleText.isNotBlank()) {
            binding.title.text = titleText
        }
        if (leftTitleText.isNotBlank()) {
            binding.titleLeftTv.text = leftTitleText
        }
        if (rightTitleText.isNotBlank()) {
            binding.titleRightTv.text = rightTitleText
        }
        if (subTitleText.isNotBlank()) {
            binding.subTitle.text = subTitleText
        }
    }
}